runtime/parachains/src/disputes.rs

```diff
@@ -100,24 +101,85 @@ impl PunishValidators for () {
 	fn punish_inconclusive(_: SessionIndex, _: impl IntoIterator<Item = ValidatorIndex>) {}
 }
 
+/// Binary discriminator to determine if the expensive signature
+/// checks are necessary.
+#[derive(Clone, Copy)]
+pub enum VerifyDisputeSignatures {
+	/// Yes, verify the signatures.
+	Yes,
+	/// No, skip the signature verification.
+	///
+	/// Only done if there exists an invariant that
+	/// can guaranteed the signature was checked before.
+	Skip,
+}
+
 /// Hook into disputes handling.
 ///
 /// Allows decoupling parachains handling from disputes so that it can
 /// potentially be disabled when instantiating a specific runtime.
-pub trait DisputesHandler<BlockNumber> {
+///
+/// Returns `Ok(())` if no duplicates were present, `Err(())` otherwise.
+///
+/// Unsorted data does not change the return value, while the node side
+/// is generally expected to pass them in sorted.
+pub trait DisputesHandler<BlockNumber: Ord> {
 	/// Whether the chain is frozen, if the chain is frozen it will not accept
 	/// any new parachain blocks for backing or inclusion.
 	fn is_frozen() -> bool;
 
-	/// Handler for filtering any dispute statements before including them as part
-	/// of inherent data. This can be useful to filter out ancient and duplicate
-	/// dispute statements.
-	fn filter_multi_dispute_data(statement_sets: &mut MultiDisputeStatementSet);
+	/// Remove dispute statement duplicates and sort the non-duplicates based on
+	/// local (lower indicies) vs remotes (higher indices) and age (older with lower indices).
+	fn deduplicate_and_sort_dispute_data(
+		statement_sets: &mut MultiDisputeStatementSet,
+	) -> Result<(), ()> {
+		use core::cmp::Ordering;
+
+		// TODO: Consider trade-of to avoid `O(n * log(n))` average lookups of `included_state`
+		// TODO: instead make a single pass and store the values.
+		let n = statement_sets.len();
+		// Sort the dispute statements according to the following prioritization:
+		//  1. Prioritize local disputes over remote disputes.
+		//  2. Prioritize older disputes over newer disputes.
+		statement_sets.sort_by(|a: &DisputeStatementSet, b: &DisputeStatementSet| {
+			let a_local_block = Self::included_state(a.session, a.candidate_hash);
+			let b_local_block = Self::included_state(b.session, b.candidate_hash);
+			match (a_local_block, b_local_block) {
+				// Prioritize local disputes over remote disputes.
+				(None, Some(_)) => Ordering::Greater,
+				(Some(_), None) => Ordering::Less,
+				// For local disputes, prioritize those that occur at an earlier height.
+				(Some(a_height), Some(b_height)) => a_height.cmp(&b_height),
+				// Prioritize earlier remote disputes using session as rough proxy.
+				(None, None) => a.session.cmp(&b.session),
```

Comment:
I think your last suggestion is the most effective. Instead of a sorted vec, we should simply traverse the statement_sets to create two pools of statements to select from.

Reply:
:+1:

Reply:
Not sure we want to skip the sorting entirely, it allows for `O(n)` deduplication.